grpc-server/src/main/scala/com/wavesplatform/events/repo/UpdatesRepoImpl.scala

```diff
@@ -173,114 +152,97 @@ class UpdatesRepoImpl(directory: String)(implicit val scheduler: Scheduler)
-                val next       = iterator.next()
-                val blockBytes = next.getValue
-                res += PBBlockchainUpdated.parseFrom(blockBytes).vanilla.get
-                goUnsafe(remaining - 1)
-              } else {
-                !iterator.hasNext
-              }
-            }
-
-            goUnsafe(LEVELDB_READ_BATCH_SIZE)
-          } catch {
-            case t: Throwable =>
-              Task.raiseError(t)
-              true
-          } finally {
-            iterator.close()
-          }
-
-          if (isLastBatch) {
-            realTimeUpdates.subscribe(realTimeUpdatesForCurrentSubscription)
-            val liquidUpdates = liquidState match {
-              case None                                     => Seq.empty
-              case Some(LiquidState(keyBlock, microBlocks)) => Seq(keyBlock) ++ microBlocks
-            }
-            (res.result() ++ liquidUpdates, None)
-          } else {
-            val result       = res.result()
-            val nextTickFrom = result.lastOption.map(_.toHeight + 1)
-            (result, nextTickFrom)
+        if (isLastBatch(data)) {
+          val liquidUpdates = liquidState match {
+            case None => Seq.empty
+            case Some(LiquidState(keyBlock, microBlocks)) =>
+              val lastBlock = data.lastOption
+              require(lastBlock.forall(keyBlock.references))
+              Seq(keyBlock) ++ microBlocks
           }
+          (data ++ liquidUpdates, None)
+        } else {
+          val nextTickFrom = data.lastOption.map(_.toHeight + 1)
+          (data, nextTickFrom)
         }
-      }
-    }
+      }(_._2.isEmpty)
 
     Observable.fromTry(height).flatMap { h =>
       if (h < fromHeight) {
         Observable.raiseError(new IllegalArgumentException("Requested start height exceeds current blockchain height"))
       } else {
-        val historical = Observable
-          .unfoldEval(fromHeight.some)(readBatch)
-          .flatMap(Observable.fromIterable)
-
-        historical ++ realTimeUpdatesForCurrentSubscription
+        def readBatchStream(from: Int): Observable[BlockchainUpdated] = Observable.defer {
+          val (data, next) = concurrent.blocking(readBatch(from))
+          Observable.fromIterable(data) ++ (next match {
+            case Some(next) =>
+              readBatchStream(next)
+
+            case None =>
+              val lastPersistentUpdate = data.lastOption
+              realTimeUpdates.dropWhile(u => !lastPersistentUpdate.forall(u.references))
+          })
+        }
+        readBatchStream(fromHeight)
       }
     }
   }
 }
 
 object UpdatesRepoImpl {
-  private val LEVELDB_READ_BATCH_SIZE = 1024
-
-  private val RANGE_REQUEST_MAX_SIZE = 1000
-  private val RANGE_REQUEST_TIMEOUT  = 1.minute
+  private val LevelDBReadBatchSize = 100
 
   private def key(height: Int): Array[Byte] =
     ByteBuffer.allocate(8).order(ByteOrder.BIG_ENDIAN).putInt(height).array()
```

Comment:
Should it be 8 bytes?

Reply:
I suppose 4 bytes would've sufficed, but oh well. I don't think now is good time to change storage structure, maybe we'll fix this in one of the future releases.